### DISCUSSION TITLE
Camera Usage for Sense Version | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#548] [Q&A]
# Camera Usage for Sense Version | Seeed Studio Wiki

This tutorial will explain how to properly use the camera on the XIAO ESP32S3 Sense.

https://wiki.seeedstudio.com/xiao_esp32s3_camera_usage/

**Comment by notevanrutten** [comment]
Will there be a guide released on how to utilize the auto-focus feature on the OV5640 module?

**Reply by xzxcessarr** [reply]
Hello, the flash bin of the module we provide include the auto-focus feature, so there is no more step for you to set.

**Reply by xzxcessarr** [reply]
If you found any problem occur, please tell us, thanks for your feedback!

**Comment by userpc42069** [comment]
Is there a way I can use this board as a webcam or a wired camera by connecting it to my laptop?

**Comment by 23Neo23** [comment]
can i use the sense and the camera to webstream using circutpython

**Comment by haya-sann** [comment]
i tried https://github.com/limengdu/SeeedStudio-XIAO-ESP32S3-Sense-camera/tree/main/CameraWebServer .
However, the Arduino IDE gave the following compiler error.
Do you have any advice on how to fix it?
```
/XIAO_ESP32S3/CameraWebServer/app_httpd.cpp: In function 'void setupLedFlash(int)':
/XIAO_ESP32S3/CameraWebServer/app_httpd.cpp:1389:5: error: 'ledcSetup' was not declared in this scope; did you mean 'ledc_stop'?
 1389 |     ledcSetup(LED_LEDC_CHANNEL, 5000, 8);
      |     ^~~~~~~~~
      |     ledc_stop
/XIAO_ESP32S3/CameraWebServer/app_httpd.cpp:1390:5: error: 'ledcAttachPin' was not declared in this scope; did you mean 'ledcAttach'?
 1390 |     ledcAttachPin(pin, LED_LEDC_CHANNEL);
      |     ^~~~~~~~~~~~~
      |     ledcAttach

exit status 1

Compilation error: 'ledcSetup' was not declared in this scope; did you mean 'ledc_stop'?
```

**Reply by aDW-code** [reply]
Hi, I researched the error message you provided and found a similar issue discussed in this [forum topic](https://forum.arduino.cc/t/ledcsetup-with-esp32-core-v3-0-0-not-work/1266221). It seems the issue stems from a change in the ESP32 core version, as the old API was removed. This [reference document](https://github.com/espressif/arduino-esp32/blob/master/docs/en/migration_guides/2.x_to_3.0.rst#ledc) provides more details on the changes. You might be able to resolve the error by updating your code according to the compiler's suggestions.

**Comment by gesGitHub** [comment]
How do we tell which version of ESP32S3 Sense or the Camera module  is in my possession?  Version 2 or version 3?

**Reply by gesGitHub** [reply]
I have v1.0 of the camera module but how to tell which version of ESP32S3 Sense I have?

**Comment by DharienDz** [comment]
Hi, I have been trying so hard to make my own camerawebser program, but I have had complications with my code, the camera and the web server do not work and apparently a watchdog is activated in the program, displaying the same messages on the serial port.

my code
#include <WiFi.h>
#include "esp_camera.h"
#include "esp_http_server.h"

// Configuración de red WiFi
const char* ssid = "TuSSID";
const char* password = "TuPassword";

// Pines de la cámara OV2640 (para Xiao ESP32S3)
#define PWDN_GPIO_NUM     -1
#define RESET_GPIO_NUM    -1
#define XCLK_GPIO_NUM      0
#define SIOD_GPIO_NUM     18
#define SIOC_GPIO_NUM     17

#define Y9_GPIO_NUM       36
#define Y8_GPIO_NUM       37
#define Y7_GPIO_NUM       38
#define Y6_GPIO_NUM       39
#define Y5_GPIO_NUM       40
#define Y4_GPIO_NUM       41
#define Y3_GPIO_NUM       42
#define Y2_GPIO_NUM       21
#define VSYNC_GPIO_NUM    5
#define HREF_GPIO_NUM     4
#define PCLK_GPIO_NUM     3

// Declaración de funciones
void startCameraServer();

void setup() {
  Serial.begin(115200);
  Serial.println("Inicializando...");

  // Configuración de la cámara
  camera_config_t config;
  config.ledc_channel = LEDC_CHANNEL_0;
  config.ledc_timer = LEDC_TIMER_0;
  config.pin_d0 = Y2_GPIO_NUM;
  config.pin_d1 = Y3_GPIO_NUM;
  config.pin_d2 = Y4_GPIO_NUM;
  config.pin_d3 = Y5_GPIO_NUM;
  config.pin_d4 = Y6_GPIO_NUM;
  config.pin_d5 = Y7_GPIO_NUM;
  config.pin_d6 = Y8_GPIO_NUM;
  config.pin_d7 = Y9_GPIO_NUM;
  config.pin_xclk = XCLK_GPIO_NUM;
  config.pin_pclk = PCLK_GPIO_NUM;
  config.pin_vsync = VSYNC_GPIO_NUM;
  config.pin_href = HREF_GPIO_NUM;
  config.pin_sccb_sda = SIOD_GPIO_NUM;
  config.pin_sccb_scl = SIOC_GPIO_NUM;
  config.pin_pwdn = PWDN_GPIO_NUM;
  config.pin_reset = RESET_GPIO_NUM;
  config.xclk_freq_hz = 20000000;
  config.pixel_format = PIXFORMAT_JPEG;

  // Tamaño de imagen y calidad
  config.frame_size = FRAMESIZE_QVGA; 
  config.jpeg_quality = 10; 
  config.fb_count = 2; 

  // Inicializar la cámara
  if (esp_camera_init(&config) != ESP_OK) {
    Serial.println("Error al inicializar la cámara. Reiniciando...");
    ESP.restart();
  }
  Serial.println("Cámara inicializada correctamente.");

  // Conectar a la red WiFi
  WiFi.begin(ssid, password);
  Serial.println("Conectando a WiFi...");
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("\nConexión WiFi establecida.");
  Serial.print("Dirección IP: ");
  Serial.println(WiFi.localIP());

  // Iniciar el servidor de la cámara
  startCameraServer();
}

void loop() {
  delay(10000); // Mantener el sistema corriendo
}

// Configuración del servidor de cámara
void startCameraServer() {
  httpd_config_t config = HTTPD_DEFAULT_CONFIG();
  config.server_port = 80;

  httpd_handle_t server = NULL;
  if (httpd_start(&server, &config) == ESP_OK) {
    httpd_uri_t uri_stream = {
      .uri      = "/",
      .method   = HTTP_GET,
      .handler  = [](httpd_req_t* req) -> esp_err_t {
        camera_fb_t* fb = esp_camera_fb_get();
        if (!fb) {
          httpd_resp_send_500(req);
          return ESP_FAIL;
        }
        httpd_resp_set_type(req, "image/jpeg");
        httpd_resp_send(req, (const char*)fb->buf, fb->len);
        esp_camera_fb_return(fb);
        return ESP_OK;
      },
      .user_ctx = NULL
    };
    httpd_register_uri_handler(server, &uri_stream);
    Serial.println("Servidor web iniciado correctamente.");
    Serial.printf("Accede a: http://%s/\n", WiFi.localIP().toString().c_str());
  } else {
    Serial.println("Error al iniciar el servidor web.");
  }
}


serial port
ESP-ROM:esp32s3-20210327
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x8 (TG1WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40376c31
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3818,len:0x109c
load:0x403c9700,len:0x4
load:0x403c9704,len:0xb50
load:0x403cc700,len:0x2fe4
entry 0x403c98ac
Inicializando...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
.......

**Comment by Mnadi121** [comment]
That is Generic ONVIF and RTSP support for legacy camera connection?

**Reply by nfs0619** [reply]
Thank you for your feedback! Currently, we support OV5640 and OV2640 cameras, and we've made some modifications to [this library](https://github.com/espressif/esp32-camera) to enable support. In principle, the ONVIF and RTSP support you mentioned can be implemented, but it will require some software modifications on your end to achieve this.

That being said, we highly recommend using the cameras we officially support for the best performance and compatibility. Feel free to try out the modifications and provide any further feedback — we'll continue to improve and enhance the support.

**Comment by skittleson** [comment]
For micropython with OV5640 cam, i had to flash it with the latest release from https://github.com/cnadler86/micropython-camera-API in releases section for this module.  Here is github gist of an example sending it to a ntfy notification https://gist.github.com/skittleson/e869a039a01b66fb8add218053b663ef

**Comment by 1Krosx1** [comment]
Is SX esp32 s3 support other 0V5460?

**Reply by nfs0619** [reply]
Hello, we recommend using the [**OV5640** camera](https://www.seeedstudio.com/OV5640-Camera-for-XIAO-ESP32S3-Sense-With-Heat-Sink-p-5739.html) **that is recommended in the documentation** together with the **XIAO ESP32S3 Sense**.
If you encounter any issues, please feel free to contact our technical support team at **[techsupport@seeed.io](mailto:techsupport@seeed.io)**.

**Comment by albinea** [comment]
i have xiao esp32s3 sense with OV3660 sensor no matter what i do i cant get video streaming and capture image only the website is loading i cant get the image

**Reply by bleaaach** [reply]
Hi,
Please provide the relevant logs, and confirm whether you have configured the PSRAM according to our wiki. Please do not skip any steps.
Please flash a model on our SenseCraft AI platform to ensure that the camera functions properly on the hardware.
https://sensecraft.seeed.cc/ai/home

**Comment by 1Krosx1** [comment]
i have Xiao and the reason why i want OV5640 is that it keeps
shattering green pigment in the original camera.

**Reply by bleaaach** [reply]
Hi, could you please provide relevant screenshots so that we can analyze the issue? Was the problem present as soon as you received the product? Additionally, do you also get the same green screen when testing on our SenseCraft AI platform?
https://sensecraft.seeed.cc/ai/home

**Comment by albinea** [comment]
i used the product as soon as i got it i tested the camera module after a
few days then i knew the thing was not working
as screen shot what should i give u
how to get the logs of this board i havent used sense craft ai i will try
it and share what happens

**Reply by nfs0619** [reply]
For your issue, please contact our technical support at [techsupport@seeed.io](mailto:techsupport@seeed.io).

**Reply by albinea** [reply]
i have an update when i change the pixel format to grayscale its working video streaming and capturing

**Comment by davefes** [comment]
In this tutorial you mention:
esp_err_t esp_camera_save_to_nvs(const char *key);
esp_err_t esp_camera_load_from_nvs(const char *key);
but I do not see them anywhere in source code.  Could you point me to 
the relevant information.

Thank you.

**Reply by nfs0619** [reply]
You can find the related functions in the following source files:

* https://github.com/espressif/arduino-esp32/blob/2.0.2/tools/sdk/esp32/include/esp32-camera/driver/include/esp_camera.h
* https://github.com/espressif/arduino-esp32/blob/caef4006af491130136b219c1205bdcf8f08bf2b/tools/sdk/esp32/include/esp32-camera/driver/include/esp_camera.h#L200

Hope this helps!

**Reply by davefes** [reply]
I have tried to use these functions and when I saw it mentioned in this tutorial I hoped to learn from your example.  I meant I could not see those functions being used in your source code.  I didn't see `Preferences.h` anywhere, so maybe you had used EEPROM to store some variables using those functions in the past.

Thank you.

**Comment by sairfan1** [comment]
I'm new to seeed studio products, I want to know if you also provide code for ESP-IDF?

**Reply by limengdu** [reply]
Our tutorials and Wiki documentation are primarily geared toward beginners, so the main development platforms we support are Arduino and PlatformIO. For ESP-IDF, we generally don't build dedicated tutorial content around it unless a specific project requires the IDF environment exclusively.

That said, Espressif has already provided excellent and comprehensive ESP-IDF documentation and development resources. Since our XIAO ESP32 series products use standard Espressif chips, you can fully rely on Espressif's official documentation for ESP-IDF development:
https://docs.espressif.com/projects/esp-idf/en/latest/

For hardware pin mapping, you can refer to the pinout diagrams on the corresponding product pages in our Wiki and map them to your ESP-IDF GPIO configurations accordingly.

**Comment by akhilgupta1002** [comment]
i want to read qr code with helop esp32s3 sense and want to display the content of qr on display .but issue is that no such library is present in arduino for reading the qr code.What should i do?
